File: leetcode/algorithms/207_course_schedule/main.cpp
#include <queue>
#include <vector>
using namespace std;

class CourseSchedule {
public:
    /**
     * Kahn's Algorithm - BFS
     *
     * Complexities:
     *   V - `numCourses`
     *   E - Size of `prerequisites`
     *   - Time Complexity: O(V + E)
     *   - Space Complexity: O(V + E)
     */
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> prerequisiteToCourse(numCourses);
        vector<int> remainedCourseCount(numCourses, 0);

        for (vector<int>& prerequisite : prerequisites) {
            int course = prerequisite[0];
            int prerequisiteForCourse = prerequisite[1];

            prerequisiteToCourse[prerequisiteForCourse].push_back(course);
            remainedCourseCount[course]++;
        }

        queue<int> courseCanTakeQueue;
        for (int course = 0; course < numCourses; course++) {
            if (remainedCourseCount[course] == 0) {
                courseCanTakeQueue.push(course);
            }
        }

        int completedCourseCount = 0;

        while (!courseCanTakeQueue.empty()) {
            int completedCourse = courseCanTakeQueue.front();
            courseCanTakeQueue.pop();
            completedCourseCount++;

            for (int course : prerequisiteToCourse[completedCourse]) {
                remainedCourseCount[course]--;

                if (remainedCourseCount[course] == 0) {
                    courseCanTakeQueue.push(course);
                }
            }
        }

        return completedCourseCount == numCourses;
    }


    // Solution
    /**
     * Kahn's Algorithm - BFS
     *
     * Complexities:
     *   V - `numCourses`
     *   E - Size of `prerequisites`
     *   - Time Complexity: O(V + E)
     *   - Space Complexity: O(V + E)
     */
    bool solution(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj(numCourses);
        vector<int> inDegree(numCourses, 0);

        for (const auto& pre : prerequisites) {
            int course = pre[0];
            int prerequisite = pre[1];
            
            adj[prerequisite].push_back(course);
            inDegree[course]++;
        }

        queue<int> q;
        for (int i = 0; i < numCourses; ++i) {
            if (inDegree[i] == 0) {
                q.push(i);
            }
        }

        int completedCount = 0;
        
        while (!q.empty()) {
            int curr = q.front();
            q.pop();
            completedCount++;

            for (int nextCourse : adj[curr]) {
                inDegree[nextCourse]--;
                
                if (inDegree[nextCourse] == 0) {
                    q.push(nextCourse);
                }
            }
        }

        return completedCount == numCourses;
    }
};

File: leetcode/algorithms/207_course_schedule/main.dart
import 'dart:collection';

class CourseSchedule {
  /**
   * Kahn's Algorithm - BFS
   *
   * Complexities:
   *   V - `numCourses`
   *   E - Size of `prerequisites`
   *   - Time Complexity: O(V + E)
   *   - Space Complexity: O(V + E)
   */
  bool canFinish(int numCourses, List<List<int>> prerequisites) {
    List<List<int>> prerequisiteToCourse = List.generate(
      numCourses,
      (index) => List<int>.empty(growable: true),
    );
    List<int> coursePrerequisiteCount = List<int>.filled(numCourses, 0);

    for (final [course, prerequisite] in prerequisites) {
      prerequisiteToCourse[prerequisite].add(course);
      coursePrerequisiteCount[course]++;
    }

    Queue<int> courseCanTakeQueue = Queue<int>();
    for (int course = 0; course < numCourses; course++) {
      if (coursePrerequisiteCount[course] == 0) {
        courseCanTakeQueue.add(course);
      }
    }

    int completedCourse = 0;

    while (courseCanTakeQueue.isNotEmpty) {
      int courseTaken = courseCanTakeQueue.removeFirst();
      completedCourse++;

      for (int course in prerequisiteToCourse[courseTaken]) {
        coursePrerequisiteCount[course]--;
        if (coursePrerequisiteCount[course] == 0) {
          courseCanTakeQueue.add(course);
        }
      }
    }

    return completedCourse == numCourses;
  }


  // Solution
  /**
   * Solution 1
   * 
   * Kahn's Algorithm - BFS
   *
   * Complexities:
   *   V - `numCourses`
   *   E - Size of `prerequisites`
   *   - Time Complexity: O(V + E)
   *   - Space Complexity: O(V + E)
   */
  bool solution1(int numCourses, List<List<int>> prerequisites) {
    List<List<int>> adj = List.generate(numCourses, (_) => []);
    List<int> inDegree = List.filled(numCourses, 0);

    for (var pre in prerequisites) {
      int course = pre[0];
      int prereq = pre[1];

      adj[prereq].add(course);
      inDegree[course]++;
    }

    Queue<int> queue = Queue();
    for (int i = 0; i < numCourses; i++) {
      if (inDegree[i] == 0) {
        queue.add(i);
      }
    }

    int takenCourses = 0;

    while (queue.isNotEmpty) {
      int curr = queue.removeFirst();
      takenCourses++;

      for (int nextCourse in adj[curr]) {
        inDegree[nextCourse]--;

        if (inDegree[nextCourse] == 0) {
          queue.add(nextCourse);
        }
      }
    }

    return takenCourses == numCourses;
  }

  /**
   * Solution 2
   * 
   * 3-Color State - DFS
   *
   * Complexities:
   *   V - `numCourses`
   *   E - Size of `prerequisites`
   *   - Time Complexity: O(V + E)
   *   - Space Complexity: O(V + E)
   */
  bool solution2(int numCourses, List<List<int>> prerequisites) {
    List<List<int>> adj = List.generate(numCourses, (_) => []);
    for (var pre in prerequisites) {
      adj[pre[1]].add(pre[0]);
    }

    List<int> state = List.filled(numCourses, 0);

    bool hasCycle(int node) {
      if (state[node] == 1) {
        return true;
      }
      if (state[node] == 2) {
        return false;
      }

      state[node] = 1;

      for (int nextNode in adj[node]) {
        if (hasCycle(nextNode)) {
          return true;
        }
      }

      state[node] = 2;
      return false;
    }

    for (int i = 0; i < numCourses; i++) {
      if (state[i] == 0) {
        if (hasCycle(i)) {
          return false;
        }
      }
    }

    return true;
  }
}
